rs/orchestrator/src/upgrade.rs

```diff
@@ -136,20 +140,69 @@ impl Upgrade {
+                    let subnet_id =
+                        get_subnet_id(&*self.registry.registry_client, cup).map_err(|err| {
+                            OrchestratorError::UpgradeError(format!(
+                                "Couldn't determine the subnet id: {:?}",
+                                err
+                            ))
+                        })?;
+                    (subnet_id, maybe_proto, maybe_cup)
+                }
+                (None, Some(proto)) => {
+                    // We found a local CUP proto that we can't deserialize. This may only happen
+                    // if this is the first CUP we are reading on a new replica version after an
+                    // upgrade. This means we have to be an assigned node, otherwise we would have
+                    // left the subnet and deleted the CUP before upgrading to this version.
+                    // The only way to leave this branch is via subnet recovery.
+
+                    // Try to find the subnet ID by deserializing only the NiDkgId. If it fails
+                    // we will have to recover using failover nodes.
+                    let nidkg_id: NiDkgId = try_from_option_field(proto.signer.clone(), "NiDkgId")
+                        .map_err(|err| {
+                            OrchestratorError::UpgradeError(format!(
+                                "Couldn't deserialize NiDkgId to determine the subnet id: {:?}",
+                                err
+                            ))
+                        })?;
+
+                    let subnet_id = match nidkg_id.target_subnet {
+                        NiDkgTargetSubnet::Local => nidkg_id.dealer_subnet,
+                        NiDkgTargetSubnet::Remote(_) => {
+                            // If this CUP was created by a remote subnet, then it is a genesis/recovery
+                            // CUP. This is the only case in the branch where we can trust the subnet ID
+                            // of the latest registry version, as switching to a registry CUP "resets" the
+                            // "oldest registry version in use" which is responsible for subnet membership.
+                            match self.registry.get_subnet_id(latest_registry_version) {
+                                Ok(subnet_id) => subnet_id,
+                                Err(OrchestratorError::NodeUnassignedError(_, _)) => {
+                                    // If the registry says that we are unassigned, this unassignment
+                                    // must have happened after the registry CUP triggering the upgrade.
+                                    // Otherwise we would have left the subnet before upgrading. This means
+                                    // we will trust the registry and go ahead with removing the node's state
+                                    // including the broken local CUP.
+                                    self.remove_state()?;
```

Comment:
Two questions: 

- At the end of this function we also stop the replica before calling `remove_state`, why don't we need this here?
- Is it clear that removing the state in this branch here is the right tradeoff? E.g. could we be better off in certain failure scenarios if we'd not touch the state here?

Reply:
> At the end of this function we also stop the replica before calling remove_state, why don't we need this here?

The local CUP cannot (fully) be deserialized so there should be no way we started a replica process with it.

> Is it clear that removing the state in this branch here is the right tradeoff? E.g. could we be better off in certain failure scenarios if we'd not touch the state here?

In my mind the risk here is similar to a recovery with failover nodes in general. If we ever recovered a subnet on an entirely new set of failover nodes, without first downloading and preserving the state somewhere, then we would loose the state of the subnet.

The situation in this branch is that we did a normal recovery with upgrade (to an incompatible version) by submitting a higher registry CUP for the same set of nodes. The nodes then derive a local CUP from the registry, persist it to disk and perform the upgrade. After they come back up they can no longer read the locally persisted registry CUP. If they now notice that they aren't part of the registry anymore according to the latest registry version, that membership change must have been adopted _after_ the recovery CUP. Otherwise the node would have left the subnet and deleted the state already before upgrading to the incompatible version.

To me this implies that the proposal for this membership change which could potentially delete the node's state must have been adopted already with the previous recovery (and its implications) in mind.

Reply:
Thanks for the explanation. Makes sense. I'm still trying to put all the pieces of the reasoning regarding when we are in which branch, based on the comments above, together and left one more question above. 

As already said earlier -- my ramp-up wrt this code here should not block progress of the PR. So if the other reviewers are confident in the change please don't block the PR on my questions, we can also discuss this offline.